SANS - automatically update default save formats

DavidFair opened the pull request:
**Description of work.**
- Automatically updates the selected save formats when toggling between 1D and 2D reduction modes

**Report to:** ISIS SANS / Whole Team

**To test:**
- Download the ISIS training course data, locate `loqdemo`
- Create a empty working folder
- In Mantid change manage user directories output directory to point to this clean directory
- Open ISIS SANS
- Set the user file to either the toml or .txt file in LOQ demo
- Set the batch file to the .csv to the file in that directory
- On the same screen try toggling between 1D and 2D, check that it updates save options each time
- Run process on the first line each time, checking the output dir contains what's expected deleting the files between
- Try toggling to memory / file / both to see that the option is respected

Fixes #31929 . <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Rename RunTabModelTest to TestRunTabModel

This brings the class under test into the same naming convention as the
file names

DavidFair committed:
Add storage to model for Reduction dim

Re #31929
Store the current reduction dimension into the model. This does not wire
up the model to change the default save yet

DavidFair committed:
Wire up observer for reduction dimension

Re #31929
Wires up an observer in the presenter for the reduction dimension
so that we can later trigger actions in the view

DavidFair committed:
Switch to GenericObserver types and PEP8 filename

Re #31929
Switches to GenericObserver types instead of creating a bespoke derived
class in each case.
Switches a non-PEP8 name to the correct style

DavidFair committed:
Wire up presenter to call new model method in callback

Re #31929
Wire up the presenter to correctly call the new model without removing
the existing logic

DavidFair committed:
Add logic to update default save opts on reduction change

Adds logic to the model to update the default save mode whenever the
reduction mode changes

DavidFair committed:
Remove logic from presenter and use new model logic

Re #31929
Removes the logic to decide whether to tick save options (which was
broken) out of presenter and use new logic in model.

Whilst the tests removed consider if we are in memory mode, we
completely ignore this now. Since we want to update the options
regardless and the model will ultimately ignore the options regardless.

This massively simplifies the logic for something minor visually.

DavidFair committed:
Wire up new reduction dim logic into view

Re #31929
Wire up the new logic into view to complete this work. This does not
remove the obselete methods yet

DavidFair committed:
Remove on_red_dim_changed from view

Re #31929

Removes old callback interface that originally was in lieu of using the
modern observer / oberservable pattern, now that it's no longer used

DavidFair committed:
Fix no save options being set by default

Re #31929
Fixes no save options appearing when the GUI starts as the model does
not start with the correct presets

DavidFair committed:
Rename run_tab_presenter_test to test_*

Re #31929

Renames run_tab_presenter_test to test_run_tab_presenter to bring it
in-line with the project standards

DavidFair committed:
Add release notes for toggle save opts

Re #31929

Adds release notes to complete the patch series

sf1919 commented:
I followed the test outlined above and all seemed to work well. However I did come across one tiny glitch.
If you do the following steps the CanSANS(1d) checkbox remains greyed out when it shouldn't:-
1) Click Memory
2) Click 2D
3) Change Memory to File

If you clicked 1D then it is still greyed out which means you still can't change it. Even if you then click both it is still greyed out.  The only way to reset it that I've found is to click on 1D, then Memory and then click on either File or Both and nothing is greyed out. Just a minor thing and I'm not sure if its really a problem CanSANS(1D) will still be ticked, even if greyed out, should that be the default/preferred option.

Does this need resolving?

sf1919 reviewed (changes requested):
I followed the test outlined above and all seemed to work well. However I did come across one tiny glitch.
If you do the following steps the CanSAS(1d) checkbox remains greyed out when it shouldn't:-

    Click Memory
    Click 2D
    Change Memory to File

If you clicked 1D then it is still greyed out which means you still can't change it. Even if you then click both it is still greyed out. The only way to reset it that I've found is to click on 1D, then Memory and then click on either File or Both and nothing is greyed out. Just a minor thing and I'm not sure if its really a problem CanSAS(1D) will still be ticked, even if greyed out, should that be the default/preferred option.

Does this need resolving?

sf1919 commented on scripts/test/SANS/gui_logic/test_run_tab_presenter.py:
```diff
@@ -629,65 +632,6 @@ def test_that_default_name_is_used_when_export_table(self):
         args = [arg for arg in self.presenter.display_save_file_box.call_args[0]]
         self.assertEqual(args, expected_args)
 
-    def test_that_canSAS_is_disabled_if_2D_reduction(self):
```
I'm trying to understand more about the changes you've implemented. From seeing that this test has been removed and from playing with the radio buttons and checkboxes it seems that you can now use canSAS even in 2D reduction. Is that ok? I admit I did try to run it and whilst  I got some files listed in Workspaces nothing saved in the folder and I got an error message. Would this error message have been prevented previously by the design that disabled canSAS in 2d reduction?

DavidFair replied:
So prior to this PR the radio buttons incorrectly always used the 2D save method (NXSCanSas) and did not automatically switch as the logic was not firing in certain cases, as it was being done in a mix of the view and presenter.

Now that the logic is correctly in the model and (checking with the SANS group) the default option which automatically toggles "should be" the they save in. As RKH is a legacy format they're trying to move away from.
Because of this I initially simplified the logic on our end, as users who are fiddling with the save options are likely instrument scientists enabling this format despite it gradually going away.

As per the bug you found above where it gets greyed out, I think I'll remove the step greying it out as it becomes overly complex as the correct approach. For example, do we untick CanSAS and grey it out whilst someone switches to 2D mode, or leave it ticked and greyed out? Do we sometimes ignore the state in our model and do something different? What happens when we switch back to 1D mode, do we restore the tick box?

Because of all of this, I decided to leave it as "what you see is what you get". As long as it gives a sane error message (which I can check) I'm okay with this as a compromise, as a user has to manually adjust the saving options to have this happen, so takes responsibility at that point.

However, again I'm open to thoughts and opinions on this?

sf1919 replied:
I think it is probably a good idea to remove the greying out as that can potentially lead to more confusion. So long as it switches to the correct tick box depending on the resolution choice then perhaps it's fair that manually changing the save type is at the user's discretion.

Could you check the error message though? I think it mentioned something about histograms but to a non-expert user like myself I'm not sure it was particularly clear it was a bad choice of save type.

sf1919 reviewed (changes requested):
I've reviewed the code now as well and have some queries that will need addressing before I can approve.

DavidFair commented on scripts/SANS/sans/gui_logic/presenter/run_tab_presenter.py:
```diff
@@ -38,7 +39,7 @@
 from sans.gui_logic.models.state_gui_model import StateGuiModel
 from sans.gui_logic.models.table_model import TableModel
-from sans.gui_logic.presenter.Observers.run_tab_observers import SaveOptionsObserver, RunTabObservers
+from sans.gui_logic.presenter.Observers.run_tab_observers import RunTabObservers
```
Good questions:

> [Replacing] SaveOptionsObserver with generic observers

This is because unlike C++, Python uses duck typing. So having a class per observer provides no real advantages. They all act and behave like generic observers in the end. Rather than having a file of stub classes I decided to just use the generic observer instead

> why continue to use RunTabObservers

`RunTabObservers` is a dataclass which simply holds all the different observer types `run_tab_presenter` can require. Ideally we would just put them into the presenter, but as our `run_tab_presenter` is effectively 10+ widgets bundled into one "god-ui" I've been trying to gradually chunk things out rather than add to it. For this reason it makes sense (to me at least) to group them together into a class since I can foresee long-term us moving more into this observer pattern to match project GUI standards.
I'm happy to shuffle the members somewhere else though, or any other ideas?

sf1919 replied:
This was more a query to understand what was going on rather than a prelude to suggested changes! It does sound like a sensible change under the circumstances and I have no idea where else these could be shuffled to or any other ideas. Need to look up duck typing....

DavidFair replied:
Duck typing is a Python concept, in a nutshell:
"If it looks like a duck, sounds, quacks and waddles like one it's probably a duck"

Well our SaveOptionsObserver looked, acted and had the API of a GenericSaveObserver so it's easier to reason with it just being a GenericSaveObserver

DavidFair committed:
Simplify logic to disable/enable checkbox for 2d reduction

Now that the GUI automatically switches reduction, unless the user
manually overrides it, we can simplify the logic.

This is because it's not obvious what the "correct" option is for 2D
reductions with CanSASNexus. Do we uncheck the tick-box? If so do we
re-check it with memory mode on when they switch to 1D or when they
switch to file mode.

As the default options (which switch automatically) are fine for 90% of
users lets leave the complexity out and let a power user make their own
decisions.

In the next commit we will improve the error message trying to process
2D data with CanSasNexus enabled

DavidFair committed:
Correctly raise when failing to save NxsCanSas

Raises rather than capturing the error if we fail to save. That way the
row correctly is marked as an error and the user doesn't blindly assume
their data is saved.

sf1919 commented:
I have tested the changes and it works well. I think the re-worded warning is more helpful. I also reviewed the changed code and it all looks good. I have therefore approved the PR.
